### PR TITLE
AMDGPU: Define v_mfma_f32_32x32x16_bf16 for gfx950

### PR DESCRIPTION
**Pull request by arsenm**
Unlike the existing gfx940 intrinsics using short/i16 in place of
bfloat, this uses the natural bfloat type.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#116680** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116680?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116679** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#116678** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116678?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116312** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116312?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116311** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116311?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116310** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116310?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116309** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116309?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116308** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116308?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116307** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116307?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#116680** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116680?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116679** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#116678** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116678?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116312** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116312?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116311** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116311?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116310** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116310?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116309** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116309?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116308** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116308?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116307** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116307?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: clang/include/clang/Basic/BuiltinsAMDGPU.def [hunk 1]
```diff
@@ -437,6 +437,8 @@ TARGET_BUILTIN(__builtin_amdgcn_cvt_sr_fp8_f32, "ifiiIi", "nc", "fp8-conversion-
 TARGET_BUILTIN(__builtin_amdgcn_mfma_f32_16x16x32_f16, "V4fV8hV8hV4fIiIiIi", "nc", "gfx950-insts")
 TARGET_BUILTIN(__builtin_amdgcn_mfma_f32_32x32x16_f16, "V16fV8hV8hV16fIiIiIi", "nc", "gfx950-insts")
```
Comment (shiltian) [inline clang/include/clang/Basic/BuiltinsAMDGPU.def:439]: nit: maybe no need for a blank line

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 2]
```diff
@@ -3117,6 +3117,8 @@ def int_amdgcn_cvt_sr_fp8_f32 : ClangBuiltin<"__builtin_amdgcn_cvt_sr_fp8_f32">,
 defset list<Intrinsic> AMDGPUMFMAIntrinsics950 = {
 def int_amdgcn_mfma_f32_16x16x32_f16 : AMDGPUMfmaIntrinsic<llvm_v4f32_ty, llvm_v8f16_ty>;
 def int_amdgcn_mfma_f32_32x32x16_f16 : AMDGPUMfmaIntrinsic<llvm_v16f32_ty, llvm_v8f16_ty>;
+
```
Comment (shiltian) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:3120]: ditto

**Review by shiltian — APPROVED** [review]

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 2 above)
Comment (arsenm) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:3120]: It’s merging the branch patch, I’m not going out of the way to introduce white space conflicts

**Comment by arsenm** [comment]
### Merge activity

* **Nov 19, 12:46 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/116679).

**Pull request merged**